bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -3251,6 +3268,41 @@ void ReplicaImp::onMessage<SimpleAckMsg>(SimpleAckMsg *msg) {
   delete msg;
 }
 
+template <>
+void ReplicaImp::onMessage<ReplicaRestartReadyMsg>(ReplicaRestartReadyMsg *msg) {
+  LOG_INFO(GL,
+           "Recieved  ReplicaRestartReadyMsg from sender_id " << std::to_string(msg->idOfGeneratedReplica())
+                                                              << " with seq_num" << std::to_string(msg->seqNum()));
+  restart_ready_list_.emplace_back(std::make_unique<ReplicaRestartReadyMsg>(msg));
+  bool bft = bftEngine::ControlStateManager::instance().getRestartBftFlag();
+  uint32_t targetNumOfMsgs = (bft ? (config_.getnumReplicas() - config_.getfVal()) : config_.getnumReplicas());
+  if (restart_ready_list_.size() == targetNumOfMsgs) {
+    LOG_INFO(GL, "Target number = " << targetNumOfMsgs << " of restart ready msgs are recieved. Send resatrt proof");
+    sendReplicasRestartReadyProof();
+  }
+}
+
+template <>
+void ReplicaImp::onMessage<ReplicasRestartReadyProofMsg>(ReplicasRestartReadyProofMsg *msg) {
+  LOG_INFO(GL,
+           "Recieved  ReplicasRestartReadyProofMsg from sender_id "
+               << std::to_string(msg->idOfGeneratedReplica()) << " with seq_num" << std::to_string(msg->seqNum()));
+  // TODO: add code call handler for restart
+}
+
+void ReplicaImp::sendReplicasRestartReadyProof() {
+  auto seq_num_to_stop_at = ControlStateManager::instance().getCheckpointToStopAt();
+  if (seq_num_to_stop_at.has_value()) {
+    unique_ptr<ReplicasRestartReadyProofMsg> restartProofMsg(
+        ReplicasRestartReadyProofMsg::create(config_.getreplicaId(), seq_num_to_stop_at.value()));
+    for (auto &msg : restart_ready_list_) {
+      restartProofMsg->addElement(msg);
+    }
+    restartProofMsg->finalizeMessage();
+    sendToAllOtherReplicas(restartProofMsg.get());
```

Comment:
See my comment above, this method doesn't send the message to the replica itself (as far as I know)